core/src/main/java/io/grpc/Context.java

```
@@ -0,0 +1,210 @@
+   */
+  public static final Context ROOT = null; // FIXME
+
+  public interface Listener {
+    public void closed(Context context);
+  }
+
+  /**
+   * Create a {@link Key} with the given name.
+   */
+  public static <T> Key<T> key(String name) {
+    return new Key<T>(name);
+  }
+
+  /**
+   * Create a {@link Key} with the given name and default value.
+   */
+  public static <T> Key<T> keyWithDefault(String name, T defaultValue) {
+    return new Key<T>(name, defaultValue);
+  }
+
+  /**
+   * Retrieve the value for the key for this context, will traverse the parent chain.
+   * @param key
+   * @return
+   */
+  // Discussion point, do we want this or do we make everyone use Key including in the builder
+  // Note that we are not providing any iteration of the values in the context.
+  public abstract <T> T get(Key<T> key);
+
+  /**
+   * Return the current context, will never return null.
+   */
+  // Discussion point. Is choosing to always return ROOT if the thread-local is empty useful?
+  public static Context current() {
+    return null;
+  }
+
+
+  /**
+   * Detach the current context from the thread and restore the context that was previously
+   * attached to the thread as the 'current' context. It is perfectly fine to attach a
+   * context which is already closed.
+   *
+   * @throws java.lang.IllegalStateException if this context is not current.
+   * @return the current context.
+   */
+  public abstract Context detach();
+
+
+  /**
+   * Attach this context to the thread and make it current, the previously current context
+   * will be restored when detach is called.
+   * @return
+   */
+  // Discussion point, make fluent?
+  public abstract void attach();
+
+
+  /**
+   * Close this context and optionally provide a cause for the closure. If this context
+   * cascades closure then notify all child contexts of closure.
+   *
+   * @throws java.lang.IllegalStateException if {@link #close()} has already been called
```

Closeable#close does not return boolean so for the moment I've chosen to silently ignore if it's already closed